crypto: Make crypto.Key implement cipher.AEAD

fd0 opened the pull request:
This was suggested by @FiloSottile in #1190

Fixes #1190

fd0 committed:
Make crypto.Key implement cipher.AEAD

fd0 committed:
Remove custom Encrypt/Decrypt methods

fd0 committed:
Run tests on Seal/Open

FiloSottile commented on internal/crypto/crypto.go:
```diff
@@ -233,79 +235,127 @@ func (k *EncryptionKey) Valid() bool {
-	if len(plaintext) < plaintextLength {
-		return 0, errors.Errorf("plaintext buffer too small, %d < %d", len(plaintext), plaintextLength)
+	if len(dst) < plaintextLength {
+		return nil, errors.Errorf("plaintext buffer too small, %d < %d", len(dst), plaintextLength)
```
Open is an append-like API. It should not require a preallocated buffer, dst can be nil.

fd0 replied:
Ah, you're right.

FiloSottile commented on internal/checker/checker.go:
```diff
@@ -724,15 +724,21 @@ func checkPack(ctx context.Context, r restic.Repository, id restic.ID) error {
+
+		nonceSize := r.Key().NonceSize()
+		nonce, ciphertext := buf[:nonceSize], buf[nonceSize:]
+		plaintext, err := r.Key().Open(ciphertext, nonce, ciphertext, nil)
```
Also, since it’s an append API it should not overwrite the slice, but add data after it. That’s why you use ciphertext[:0].

fd0 replied:
Ok, need to make that sure. Thanks!

FiloSottile reviewed (changes requested):
I had a very quick look, and it’s a good change, but I think the API is a little off.

fd0 commented:
Thanks a lot for your comments, I'll fix it.

fd0 committed:
Use Seal/Open everywhere

fd0 commented:
Thank you very much for your comments, I indeed misunderstood the API. I've corrected the implementation, and added tests for the corner cases.

Please take another look!

FiloSottile commented on internal/repository/key.go:
```diff
@@ -87,12 +87,12 @@ func OpenKey(ctx context.Context, s *Repository, name string, password string) (
-	n, err := k.user.Decrypt(buf, k.Data)
+	nonce, ciphertext := k.Data[:k.user.NonceSize()], k.Data[k.user.NonceSize():]
+	buf := make([]byte, 0, len(ciphertext))
+	buf, err = k.user.Open(buf, nonce, ciphertext, nil)
```
Here and elsewhere you could skip the `make([]byte, 0, len(ciphertext))` and pass nil instead, as that will be done automatically.

fd0 replied:
Good point, although I did only find this one occurrence. I've replaced it.

FiloSottile commented on internal/crypto/crypto.go:
```diff
@@ -233,91 +235,142 @@ func (k *EncryptionKey) Valid() bool {
+func validNonce(nonce []byte) bool {
+	sum := 0
+	for b := range nonce {
+		sum += b
```
This might be faster with `|=` and `var sum byte`.

fd0 replied:
You're right, it's ever so slightly faster, at least on amd64:
```
$ benchcmp bench-int bench-byte
benchmark                 old ns/op     new ns/op     delta
BenchmarkNonceValid-4     9.45          8.22          -13.02%
```

But, much more importantly: After I made `sum` a `byte`, the compiler suddenly complained that `b` had the wrong type: `int` instead of `byte`. Turns out, the function has a very subtle bug: It didn't sum up the elements of `nonce`, but the indexes. :disappointed: 

So, thanks again for the review, it helped find a bug! (And I've also added tests for thatt).

FiloSottile replied:
Woo, good catch!

FiloSottile commented on internal/crypto/crypto.go:
```diff
@@ -233,91 +235,142 @@ func (k *EncryptionKey) Valid() bool {
+		dst = append(dst, make([]byte, ciphertextLength-capacity)...)
+	} else {
+		dst = dst[:pos+ciphertextLength]
+	}
```
The common pattern here is `sliceForAppend`.

https://github.com/golang/go/blob/964639cc338db650ccadeafb7424bc8ebb2c0f6c/src/crypto/aes/aes_gcm.go#L81-L94

fd0 replied:
Oh, I wasn't aware of that. Thanks for the pointer! The pattern is really nice and reduces the complexity of the code in Seal/Open :)

fd0 committed:
crypto: Fix nonce test, make it faster

fd0 committed:
crypto: Correct test function names

fd0 committed:
crypto: Rework Seal/Open to use sliceForAppend

fd0 committed:
Pass in a nil buffer to Open()

fd0 commented:
Thank you very much for your help @FiloSottile!
